File: src/harness_core/dispatch/health_checker.py
"""Phase 6A: HealthChecker — storage and service health probes."""

from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Any


HEALTH_CHECKER_SCHEMA_VERSION = "health_checker.v1"


@dataclass(frozen=True)
class HealthCheck:
    name: str
    status: str
    message: str = ""
    latency_ms: float = 0.0


@dataclass(frozen=True)
class HealthReport:
    status: str
    checks: list[HealthCheck]
    timestamp: float = field(default_factory=time.time)


class HealthChecker:
    """Probes storage, events, and plan stores for health/readiness."""

    def __init__(self, store: Any = None) -> None:
        self._store = store

    def check_storage(self) -> HealthCheck:
        if self._store is None:
            return HealthCheck(name="storage", status="unhealthy", message="no store configured")
        start = time.monotonic()
        try:
            stats = self._store.stats()
            latency = (time.monotonic() - start) * 1000
            return HealthCheck(name="storage", status="healthy",
                               message=f"plans={stats['plans']}, repos={stats['repos']}, events={stats['events']}",
                               latency_ms=latency)
        except Exception as e:
            latency = (time.monotonic() - start) * 1000
            return HealthCheck(name="storage", status="unhealthy", message=str(e), latency_ms=latency)

    def check_events(self) -> HealthCheck:
        if self._store is None:
            return HealthCheck(name="events", status="unhealthy", message="no store configured")
        start = time.monotonic()
        try:
            events = self._store.get_events(limit=1)
            latency = (time.monotonic() - start) * 1000
            return HealthCheck(name="events", status="healthy",
                               message=f"accessible, latest_count={len(events)}",
                               latency_ms=latency)
        except Exception as e:
            latency = (time.monotonic() - start) * 1000
            return HealthCheck(name="events", status="unhealthy", message=str(e), latency_ms=latency)

    def check_plans(self) -> HealthCheck:
        if self._store is None:
            return HealthCheck(name="plans", status="unhealthy", message="no store configured")
        start = time.monotonic()
        try:
            plans = self._store.list_plans()
            latency = (time.monotonic() - start) * 1000
            return HealthCheck(name="plans", status="healthy",
                               message=f"accessible, count={len(plans)}",
                               latency_ms=latency)
        except Exception as e:
            latency = (time.monotonic() - start) * 1000
            return HealthCheck(name="plans", status="unhealthy", message=str(e), latency_ms=latency)

    def health(self) -> HealthReport:
        checks = [
            self.check_storage(),
            self.check_events(),
            self.check_plans(),
        ]
        statuses = [c.status for c in checks]
        if all(s == "healthy" for s in statuses):
            overall = "healthy"
        elif any(s == "unhealthy" for s in statuses):
            overall = "unhealthy"
        else:
            overall = "degraded"
        return HealthReport(status=overall, checks=checks)

    def readiness(self) -> HealthReport:
        checks = [
            self.check_storage(),
            self.check_events(),
            self.check_plans(),
        ]
        ready = all(c.status == "healthy" for c in checks)
        return HealthReport(
            status="ready" if ready else "not_ready",
            checks=checks,
        )

    def health_dict(self) -> dict[str, Any]:
        report = self.health()
        return {
            "status": report.status,
            "checks": {c.name: {"status": c.status, "message": c.message, "latency_ms": c.latency_ms}
                       for c in report.checks},
            "timestamp": report.timestamp,
        }

    def readiness_dict(self) -> dict[str, Any]:
        report = self.readiness()
        return {
            "ready": report.status == "ready",
            "status": report.status,
            "checks": {c.name: {"status": c.status, "message": c.message}
                       for c in report.checks},
        }

File: engine/src/storage/health_checker.rs
use super::durable_store::DurableStore;
use serde::{Deserialize, Serialize};

pub const HEALTH_CHECKER_SCHEMA_VERSION: &str = "health_checker.v1";

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HealthCheck {
    pub name: String,
    pub status: String,
    pub message: String,
    pub latency_ms: f64,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct HealthReport {
    pub status: String,
    pub checks: Vec<HealthCheck>,
    pub timestamp: f64,
}

pub struct HealthChecker<'a> {
    store: Option<&'a DurableStore>,
}

impl<'a> HealthChecker<'a> {
    pub fn new(store: Option<&'a DurableStore>) -> Self {
        Self { store }
    }

    pub fn check_storage(&self, now: f64) -> HealthCheck {
        match self.store {
            None => HealthCheck {
                name: "storage".to_string(),
                status: "unhealthy".to_string(),
                message: "no store configured".to_string(),
                latency_ms: 0.0,
            },
            Some(store) => match store.stats() {
                Ok(stats) => HealthCheck {
                    name: "storage".to_string(),
                    status: "healthy".to_string(),
                    message: format!(
                        "plans={}, repos={}, events={}",
                        stats["plans"], stats["repos"], stats["events"]
                    ),
                    latency_ms: now * 1000.0, // simplified latency
                },
                Err(e) => HealthCheck {
                    name: "storage".to_string(),
                    status: "unhealthy".to_string(),
                    message: e,
                    latency_ms: 0.0,
                },
            },
        }
    }

    pub fn check_events(&self, now: f64) -> HealthCheck {
        match self.store {
            None => HealthCheck {
                name: "events".to_string(),
                status: "unhealthy".to_string(),
                message: "no store configured".to_string(),
                latency_ms: 0.0,
            },
            Some(store) => match store.get_events(None, 1) {
                Ok(events) => HealthCheck {
                    name: "events".to_string(),
                    status: "healthy".to_string(),
                    message: format!("accessible, latest_count={}", events.len()),
                    latency_ms: now * 1000.0,
                },
                Err(e) => HealthCheck {
                    name: "events".to_string(),
                    status: "unhealthy".to_string(),
                    message: e,
                    latency_ms: 0.0,
                },
            },
        }
    }

    pub fn check_plans(&self, now: f64) -> HealthCheck {
        match self.store {
            None => HealthCheck {
                name: "plans".to_string(),
                status: "unhealthy".to_string(),
                message: "no store configured".to_string(),
                latency_ms: 0.0,
            },
            Some(store) => match store.stats() {
                Ok(stats) => HealthCheck {
                    name: "plans".to_string(),
                    status: "healthy".to_string(),
                    message: format!("accessible, count={}", stats["plans"]),
                    latency_ms: now * 1000.0,
                },
                Err(e) => HealthCheck {
                    name: "plans".to_string(),
                    status: "unhealthy".to_string(),
                    message: e,
                    latency_ms: 0.0,
                },
            },
        }
    }

    pub fn health(&self, now: f64) -> HealthReport {
        let checks = vec![
            self.check_storage(now),
            self.check_events(now),
            self.check_plans(now),
        ];
        let statuses: Vec<&str> = checks.iter().map(|c| c.status.as_str()).collect();
        let overall = if statuses.iter().all(|s| *s == "healthy") {
            "healthy"
        } else if statuses.contains(&"unhealthy") {
            "unhealthy"
        } else {
            "degraded"
        };
        HealthReport {
            status: overall.to_string(),
            checks,
            timestamp: now,
        }
    }

    pub fn readiness(&self, now: f64) -> HealthReport {
        let checks = vec![
            self.check_storage(now),
            self.check_events(now),
            self.check_plans(now),
        ];
        let ready = checks.iter().all(|c| c.status == "healthy");
        HealthReport {
            status: if ready { "ready" } else { "not_ready" }.to_string(),
            checks,
            timestamp: now,
        }
    }
}
